external_search.py

```diff
@@ -458,58 +462,93 @@ def query_works(self, query_string, filter=None, pagination=None,
             fields known to be used by the feed generation code.  This
             all comes at a slight performance cost.
         :return: A list of Hit objects containing information about
-            the search results, including the values of any script fields
-            calculated by ElasticSearch during the search process.
+            the search results. This will include the values of any
+            script fields calculated by ElasticSearch during the
+            search process.
         """
-        if not self.works_alias:
+        if isinstance(filter, Filter) and filter.match_nothing is True:
+            # We already know this search should match nothing.  We
+            # don't even need to perform the search.
             return []
 
-        if not pagination:
-            pagination = Pagination.default()
+        pagination = pagination or Pagination.default()
+        query_data = (query_string, filter, pagination)
+        [result] = list(self.query_works_multi([query_data], debug))
+        return result
 
-        search = self.create_search_doc(query_string, filter=filter, pagination=pagination, debug=debug)
-        if filter is not None and filter.match_nothing is True:
-            # We already know that the search should match nothing.
-            # We don't even need to perform the search.
-            return []
-        start = pagination.offset
-        stop = start + pagination.size
-
-        function_scores = filter.scoring_functions if filter else None
-        if function_scores:
-            function_score = FunctionScore(
-                query=dict(match_all=dict()),
-                functions=function_scores,
-                score_mode="sum"
-            )
-            search = search.query(function_score)
-        a = time.time()
+    def query_works_multi(self, queries, debug=False):
+        """Run several queries simultaneously and return the results
+        as a big list.
+
+        :param queries: A list of (query string, Filter, Pagination) 3-tuples,
+            each representing an Elasticsearch query to be run.
 
-        results = search[start:stop]
-        # Convert the Search object into a list of hits.
+        :yield: A sequence of lists, one per item in `queries`,
+            each containing the search results from that
+            (query string, Filter, Pagination) 3-tuple.
+        """
+        # If the works alias is not set, all queries return empty.
         #
+        # TODO: Maybe an unset works_alias should raise
+        # CannotLoadConfiguration in the constructor. Then we wouldn't
+        # have to worry about this.
+        if not self.works_alias:
+            for q in queries:
+                yield []
+
+        # Create a MultiSearch. 
+        multi = MultiSearch(using=self.__client)
```

Comment:
`__client` is the ElasticSearch client, right?

Reply:
Yes, it's the Elasticsearch object from elasticsearch-dsl.